gpiozero/pins/data.py

```diff
@@ -422,13 +782,185 @@ class PiBoardInfo(namedtuple('PiBoardInfo', (
+        string. For example::
+
+            from gpiozero import *
+            print('{0:board}'.format(pi_info()))
 
     .. _system on a chip: https://en.wikipedia.org/wiki/System_on_a_chip
     """
+    __slots__ = () # workaround python issue #24931
+
+    @classmethod
+    def from_revision(cls, revision):
+        if revision & 0x800000:
+            # New-style revision, parse information from bit-pattern:
+            #
+            # MSB -----------------------> LSB
+            # uuuuuuuuFMMMCCCCPPPPTTTTTTTTRRRR
+            #
+            # uuuuuuuu - Unused
+            # F        - New flag (1=valid new-style revision, 0=old-style)
+            # MMM      - Memory size (0=256, 1=512, 2=1024)
+            # CCCC     - Manufacturer (0=Sony, 1=Egoman, 2=Embest)
+            # PPPP     - Processor (0=2835, 1=2836, 2=2837)
+            # TTTTTTTT - Type (0=A, 1=B, 2=A+, 3=B+, 4=2B, 5=Alpha (??), 6=CM, 8=3B, 9=Zero)
+            # RRRR     - Revision (0, 1, or 2)
+            try:
+                model = {
+                    0: 'A',
+                    1: 'B',
+                    2: 'A+',
+                    3: 'B+',
+                    4: '2B',
+                    6: 'CM',
+                    8: '3B',
+                    9: 'Zero',
+                    }[(revision & 0xff0) >> 4]
+                if model in ('A', 'B'):
+                    pcb_revision = {
+                        0: '1.0', # is this right?
+                        1: '1.0',
+                        2: '2.0',
+                        }[revision & 0x0f]
+                else:
+                    pcb_revision = '1.%d' % (revision & 0x0f)
+                released = {
+                    'A':    '2013Q1',
+                    'B':    '2012Q1' if pcb_revision == '1.0' else '2012Q4',
+                    'A+':   '2014Q4',
+                    'B+':   '2014Q3',
+                    '2B':   '2015Q1',
+                    'CM':   '2014Q2',
+                    '3B':   '2016Q1',
+                    'Zero': '2015Q4' if pcb_revision == '1.2' else '2016Q2',
+                    }[model]
+                soc = {
+                    0: 'BCM2835',
+                    1: 'BCM2836',
+                    2: 'BCM2837',
+                    }[(revision & 0xf000) >> 12]
+                manufacturer = {
+                    0: 'Sony',
+                    1: 'Egoman',
+                    2: 'Embest',
+                    }[(revision & 0xf0000) >> 16]
+                memory = {
+                    0: 256,
+                    1: 512,
+                    2: 1024,
+                    }[(revision & 0x700000) >> 20]
+                storage = {
+                    'A': 'SD',
+                    'B': 'SD',
+                    'CM': 'eMMC',
+                    }.get(model, 'MicroSD')
+                usb = {
+                    'A':    1,
+                    'A+':   1,
+                    'Zero': 1,
+                    'B':    2,
+                    'CM':   0,
```

Comment:
Ah, good point - I was wondering the other day what on earth USB DM was and whether I'd done something wrong when copy'n'pasting the diagram. Nope - it's DM on the diagram too - so they've used N for NEGATIVE everywhere else but on the USB pins it's M for MINUS? (and therefore presumably P for PLUS instead of POSITIVE :)